lib/Exporter/ONNXModelWriter.cpp

```diff
@@ -0,0 +1,1042 @@
+          continue;
+        }
+
+        // Handle placeholders cases.
+        if (kind == Kinded::Kind::PlaceholderKind) {
+          const auto *PH = llvm::cast<Placeholder>(N);
+          // Write global input, output only tensor shape.
+          auto *inputProto = graphProto->add_input();
+          tensorShapeFromPlaceholder(PH, inputProto);
+        } else {
+          // Write global initializer, output tensor bytes.
+          const auto *C = llvm::cast<Constant>(N);
+          auto *tensorProto = graphProto->add_initializer();
+          tensorProto->set_name(C->getName());
+          writeTensor(C->getPayload(), tensorProto);
+        }
+      } else if (kind == Kinded::Kind::SaveNodeKind) {
+        // Save node case, find input and use its name as a global output,
+        // output only shape.
+        const auto *PH = llvm::cast<SaveNode>(N)->getPlaceholder();
+        auto *outputProto = graphProto->add_output();
+        tensorShapeFromPlaceholder(PH, outputProto);
+      } else {
+        RETURN_IF_ERR(writeOperator(N, *graphProto, reporter));
+      }
+      reporter(N);
+    }
+    return writeModel(modelProto, textMode);
+  };
+
+  if (errPtr) {
+    *errPtr = setup();
+  } else {
+    EXIT_ON_ERR(setup());
+  }
+}
+
+ONNXModelWriter::TensorType::DataType
+ONNXModelWriter::convertType(const Type &glowType) {
+  switch (glowType.getElementType()) {
+  case ElemKind::FloatTy:
+    return TensorType::FLOAT;
+  case ElemKind::Float16Ty:
+    return TensorType::FLOAT16;
+  case ElemKind::Int8QTy:
+    return TensorType::INT8;
+  case ElemKind::UInt8FusedQTy:
+  case ElemKind::UInt8QTy:
+    return TensorType::UINT8;
+  case ElemKind::Int16QTy:
+    return TensorType::INT16;
+  case ElemKind::Int32QTy:
+  case ElemKind::Int32ITy:
+    return TensorType::INT32;
+  case ElemKind::Int64ITy:
+    return TensorType::INT64;
+  case ElemKind::BoolTy:
+    return TensorType::BOOL;
+  }
+}
+
+void ONNXModelWriter::writeTensor(const Tensor &T, TensorType *out) {
+  const auto &type = T.getType();
+  out->set_data_type(convertType(type));
+  const auto &dims = type.dims();
+  for (unsigned b = 0, e = dims.size(); b < e; ++b) {
+    out->add_dims(dims[b]);
+  }
+
+  out->set_raw_data(std::string(T.getUnsafePtr(), type.getSizeInBytes()));
+}
+
+void ONNXModelWriter::tensorShapeFromPlaceholder(const Placeholder *PH,
+                                                 ValueInfoType *valueProto) {
+  valueProto->set_name(PH->getName());
+  auto *type = valueProto->mutable_type();
+  auto *tensorType = type->mutable_tensor_type();
+  auto *glowType = PH->getType();
+  tensorType->set_elem_type(convertType(*glowType));
+  auto *tensorShape = tensorType->mutable_shape();
+  const auto &dims = glowType->dims();
+  for (unsigned b = 0, e = dims.size(); b < e; ++b) {
+    auto *tensorDims = tensorShape->add_dim();
+    tensorDims->set_dim_value(dims[b]);
+  }
+}
+
+bool ONNXModelWriter::isIgnoreNode(const Node *node) {
+  switch (node->getKind()) {
```

Comment:
Why do we ignore these? Is it because we expect to fold them into other Nodes, e.g. folding Broadcasted inputs into Arithmetic nodes? If yes, then won't we do this folding before we need to process the Tile or Splat Node, since we do post order traversal? It seems like we should be able to fall back to exporting these, no?

In case we do keep ignoring these, both of the Nodes here have logic defined for writing (e.g. `writeSplat()`/`writeTIle()`). If we are always going to ignore them then perhaps those functions should instead throw an Error, since we never expect to write them out. Also we should document exactly why we are ignoring these nodes.

Reply:
yes, you are right. Cause the graph navigation is done in post order all cases when Tile/Splat nodes had been added can be safely tracked, but explicit Tile/Splat nodes would be unaffected.